src/metabase/query_analysis/native_query_analyzer.clj

```diff
@@ -177,17 +177,39 @@
                          [:= :t.db_id db-id]
                          (into [:or] (map table-query tables))]}))))
 
+(defn- fill-missing-table-ids-hack
+  "See if we can qualify the schema and table-id for any explicit field refs which couldn't resolve their field"
+  [table-refs field-refs]
+  ;; Note, at this point we have given up on any pretense of respecting case sensitivity
+  (let [t->ids  (group-by (comp u/lower-case-en :table) table-refs)
+        s+t->id (u/index-by (juxt (comp u/lower-case-en :schema)
+                                  (comp u/lower-case-en :table))
+                            table-refs)]
+    (into (empty field-refs)
+          (mapcat (fn [{:keys [schema table table-id] :as ref}]
+                    (if table-id
+                      [ref]
+                      (or (if schema
+                            (when-let [table-ref (s+t->id [(u/lower-case-en schema)
+                                                           (u/lower-case-en table)])]
+                              [(merge ref table-ref)])
+                            (seq (map #(merge ref %) (t->ids (u/lower-case-en table)))))
+                          [ref]))))
```

Comment:
Do you think that's equivalent? Trying to make logic a bit more flat so it'll be easier to read it.

```suggestion
                    (or
                      (when table-id [ref])
                      (when schema (when-let [table-ref (s+t->id [(u/lower-case-en schema)
                                                                  (u/lower-case-en table)])]
                                     [(merge ref table-ref)]))
                      (seq (map #(merge ref %) (t->ids (u/lower-case-en table))))
                      [ref])
```

Reply:
Yeah, this is a bit gnarly. Flattening is nice, but I like repeating inverse conditions. Will see if it can be cleaned up even further.

Reply:
Went in a [slightly different direction](https://github.com/metabase/metabase/pull/46185/commits/e57305642d96faa3998b2b0751d300f21bb560f1) - what do you think?

Reply:
Oh damn, fixed alignment in a later commit 😆

Reply:
[Better link](https://github.com/metabase/metabase/pull/46185/commits/90621b5a05ded7f0d5b1e4ef691860220ff43a01#diff-b569425886916fa8d6aa396ef80055206af8ffcb1e9cdfa37fb78988b43d7563L184-R199)